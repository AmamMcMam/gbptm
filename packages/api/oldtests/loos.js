/* global describe, before, after, it */
var config = require('../../config/config');
var baseUrl = 'http://localhost:' + config.app.port;
var co = require('co');
var supertest = require('supertest');
var request = supertest(baseUrl);
var Loo = require('../../models/loo');
var loader = require('../loader.js').dataLoader;

describe('Find loos within a box (/loos/in)', function() {
  before(function(done) {
    loader(Loo, 'looBox', function(err, result) {
      done();
      return result;
    });
  });
  after(function(done) {
    co(function*() {
      yield Loo.remove({});
    }).then(done);
  });

  it('/loos/in/:sw/:ne/:nw/:se', function(done) {
    request
      .get('/loos/in/-24.2,44.5/20.3,60.4/-24.2,60.4/20.3,44.5')
      .set('Accept', 'application/json')
      .expect(200)
      .expect(function(res) {
        if (!(res.body.features.length === 12)) {
          return 'Not enough Loos';
        }
      })
      .end(done);
  });
});

describe('Find loos within radius (/loos/near)', function() {
  // Bring up a server before testing
  before(function(done) {
    loader(Loo, 'looRadius', function(err, result) {
      done();
      return result;
    });
  });
  // tear it down after
  after(function(done) {
    co(function*() {
      yield Loo.remove({});
    }).then(done);
  });

  it('/loos/near/:lon/:lat JSON with default limit', function(done) {
    request
      .get('/loos/near/-0.2068223/51.518342/')
      .set('Accept', 'application/json')
      .query({ radius: 2000 })
      .expect(200)
      .expect(function(res) {
        if (!(res.body.features.length === 5)) {
          return 'Not correct amount of loos';
        }
      })
      .end(done);
  });

  it('/loos/near/:lon/:lat JSON with limit of 17', function(done) {
    request
      .get('/loos/near/-0.2068223/51.518342/?radius=2000&limit=20')
      .set('Accept', 'application/json')
      .expect(200)
      .expect(function(res) {
        if (!(res.body.features.length === 20)) {
          return 'Not correct amount of loos';
        }
      })
      .end(done);
  });
});

describe('Find loos via ID ( /loos/:id/)', function() {
  var looGlobal = null;
  before(function(done) {
    loader(Loo, 'looID', function(err, result) {
      looGlobal = result.ops[0];
      done();
      return result;
    });
  });

  // tear it dewn after
  after(function(done) {
    co(function*() {
      yield Loo.remove({});
    }).then(done);
  });

  it('/loos/:id JSON', function(done) {
    request
      .get('/loos/' + looGlobal._id.toString())
      .set('Accept', 'application/json')
      .expect(200)
      .expect(function(res) {
        if (!res.body) {
          return 'Not enough Loos';
        }
      })
      .end(done);
  });
});
