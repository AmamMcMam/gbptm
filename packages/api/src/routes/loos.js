const express = require('express');
const router = express.Router();
const config = require('../config/config');
const { Loo } = require('../db')(config.mongo.url);

/**
 * Get loos near lon/lat
 * Accepts `radius` in meters in query
 */
router.get('/near/:lon/:lat', async (req, res) => {
  const wantedRadius =
    parseFloat(req.query.radius) || config.query_defaults.defaultRadius;
  const allowedRadius = Math.min(wantedRadius, config.query_defaults.maxRadius);
  const loos = await Loo.findNear(
    parseFloat(req.params.lon),
    parseFloat(req.params.lat),
    allowedRadius
  ).exec();

  res.status(200).json(loos);
});

/**
 * GET an array of points
 */
router.get('/', async (req, res) => {
  const loos = await Loo.find({}, 'type geometry').exec();
  res.status(200).json(loos);
});

/**
 * Regenerate a loo from its reports
 * this'll update it's admin geo
 */
router.get('/:id/updateArea', async (req, res) => {
  const loo = await Loo.findById(req.params.id);
  const regen = await loo.regenerate();
  await regen.save();
  res.status(200).json(regen);
});

/**
 * GET a loo by ID
 */
router.get('/:id', async (req, res) => {
  const loo = await Loo.findById(req.params.id).exec();
  console.log(loo);
  if (!loo) {
    return res.status(404).end();
  }

  if (req.query.populateReports) {
    const populatedReportsLoo = await loo.populate('reports').execPopulate();
    res.status(200).json(populatedReportsLoo);
    return;
  }

  res.status(200).json(loo);
});

module.exports = router;
