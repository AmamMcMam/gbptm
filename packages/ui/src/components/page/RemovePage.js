import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { browserHistory } from 'react-router';
import { connect } from 'react-redux';

import layout from '../css/layout.module.css';
import headings from '../../css/headings.module.css';
import controls from '../../css/controls.module.css';

import config from '../../config';

class RemovePage extends Component {
  render() {
    var endpoint = `${config.apiEndpoint}/remove/${this.props.loo._id}`;

    return (
      <div>
        {this.props.app.canGoBack && (
          <div>
            <div className={layout.controls}>
              <button onClick={browserHistory.goBack} className={controls.btn}>
                Back
              </button>
            </div>
          </div>
        )}

        <h2 className={headings.large}>Toilet Remover</h2>

        <p>
          Please let us know why you're removing this toilet from the map using
          the form below.
        </p>

        <form action={endpoint} method="post">
          <label>
            Reason for removal
            <textarea type="text" name="reason" className={controls.text} />
          </label>

          <button type="submit" className={controls.btnCaution}>
            Remove it
          </button>
        </form>
      </div>
    );
  }
}

RemovePage.propTypes = {
  loo: PropTypes.object.isRequired,
};

var mapStateToProps = state => ({
  app: state.app,
});

var mapDispatchToProps = {};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(RemovePage);
