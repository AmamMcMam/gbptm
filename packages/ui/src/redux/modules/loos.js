export const FIND_NEARBY_REQUEST = 'LOOS/FIND_NEARBY/REQUEST';
export const FIND_NEARBY_SUCCESS = 'LOOS/FIND_NEARBY/SUCCESS';
export const FIND_BY_ID_REQUEST = 'LOOS/FIND_BY_ID_REQUEST/SUCCESS';
export const FIND_BY_ID_SUCCESS = 'LOOS/FIND_BY_ID_SUCCESS/SUCCESS';
export const REPORT_REQUEST = 'LOOS/REPORT/REQUEST';
export const REPORT_SUCCESS = 'LOOS/REPORT/SUCCESS';
export const REPORT_PROCESS_PENDING = 'LOOS/REPORT_PROCESS_PENDING';

// Actions

export const actionFindNearbyRequest = (lng, lat, radius) => ({
  type: FIND_NEARBY_REQUEST,
  payload: {
    lng,
    lat,
    radius,
  },
});

export const actionFindNearbySuccess = loos => ({
  type: FIND_NEARBY_SUCCESS,
  payload: {
    loos,
  },
});

export const actionFindByIdRequest = id => ({
  type: FIND_BY_ID_REQUEST,
  payload: {
    id,
  },
});

export const actionFindByIdSuccess = loo => ({
  type: FIND_BY_ID_SUCCESS,
  payload: {
    loo,
  },
});

export const actionReportRequest = loo => ({
  type: REPORT_REQUEST,
  payload: {
    loo,
  },
});

export const actionReportSuccess = () => ({
  type: REPORT_SUCCESS,
});

export const actionProcessPendingReports = () => ({
  type: REPORT_PROCESS_PENDING,
});

export const actions = {
  [FIND_NEARBY_REQUEST]: actionFindNearbyRequest,
  [FIND_NEARBY_SUCCESS]: actionFindNearbySuccess,
  [FIND_BY_ID_REQUEST]: actionFindByIdRequest,
  [FIND_BY_ID_SUCCESS]: actionFindByIdSuccess,
  [REPORT_REQUEST]: actionReportRequest,
  [REPORT_SUCCESS]: actionReportSuccess,
  [REPORT_PROCESS_PENDING]: actionProcessPendingReports,
};

// Action Handlers

const ACTION_HANDLERS = {
  [FIND_BY_ID_REQUEST]: function(state, action) {
    // Clear the stored loo reference on each request to ensure we
    // don't render against a cached instance
    return Object.assign({}, state, {
      byId: null,
    });
  },

  [FIND_NEARBY_SUCCESS]: function(state, action) {
    return Object.assign({}, state, {
      nearby: action.payload.loos || [],
    });
  },

  [FIND_BY_ID_SUCCESS]: function(state, action) {
    return Object.assign({}, state, {
      byId: action.payload.loo,
    });
  },
};

// Reducer

export default function loosReducer(state = {}, action) {
  const handler = ACTION_HANDLERS[action.type];

  return handler ? handler(state, action) : state;
}
