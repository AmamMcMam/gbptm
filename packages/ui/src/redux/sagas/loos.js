import { all, takeLatest, call, put } from 'redux-saga/effects';
import { push as navigate } from 'react-router-redux';

import api from '../../api';
import { PENDING_REPORT_KEY } from '../../config';

import {
  FIND_NEARBY_REQUEST,
  FIND_NEARBY_SUCCESS,
  FIND_BY_ID_REQUEST,
  FIND_BY_ID_SUCCESS,
  REPORT_REQUEST,
  REPORT_SUCCESS,
  REPORT_PROCESS_PENDING,
  actions,
} from '../modules/loos';

function* findNearbyLoosSaga(action) {
  var { lng, lat, radius } = action.payload;
  var loos = yield call(api.findLoos, lng, lat, radius);
  yield put(actions[FIND_NEARBY_SUCCESS](loos));
}

function* findLooByIdSaga(action) {
  var loo = yield call(api.findLooById, action.payload.id);
  yield put(actions[FIND_BY_ID_SUCCESS](loo));
}

function* report(loo) {
  // Re-direct the user to the authentication screen, if they're not
  // already logged in
  var authenticated = yield api.isAuthenticated();

  if (!authenticated) {
    yield put(navigate('/signin'));
  } else {
    // Todo: Catch HTTP 401 and navigate to '/signin'
    yield call(api.reportLoo, loo);
    yield put(actions[REPORT_SUCCESS](loo));
  }
}

function* reportRequest(action) {
  var loo = action.payload.loo;

  // Store loo in localStorage so, if neccessary, we can perform
  // authentication and pick it up again once we've been re-directed
  // back to our application
  localStorage.setItem(PENDING_REPORT_KEY, JSON.stringify(loo));

  yield call(report, loo);
}

function* processPending() {
  var loo = JSON.parse(localStorage.getItem(PENDING_REPORT_KEY));
  var authenticated = yield api.isAuthenticated();

  if (loo && authenticated) {
    yield call(api.reportLoo, loo);
    yield put(actions[REPORT_SUCCESS](loo));

    // Remove pending loo report
    localStorage.removeItem(PENDING_REPORT_KEY);
  }
}

export default function* loosSaga() {
  yield all([
    takeLatest(FIND_NEARBY_REQUEST, findNearbyLoosSaga),
    takeLatest(FIND_BY_ID_REQUEST, findLooByIdSaga),
    takeLatest(REPORT_REQUEST, reportRequest),
    takeLatest(REPORT_PROCESS_PENDING, processPending),
  ]);
}
