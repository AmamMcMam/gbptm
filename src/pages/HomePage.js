import React from 'react';
import PropTypes from 'prop-types';
import { useParams, useRouteMatch } from 'react-router-dom';
import { loader } from 'graphql.macro';
import { useQuery } from '@apollo/client';

import config from '../config';
import PageLayout from '../components/PageLayout';
import LooMap from '../components/LooMap';
import useMapPosition from '../components/useMapPosition';
import useNearbyLoos from '../components/useNearbyLoos';
import Box from '../components/Box';
import ToiletDetailsPanel from '../components/ToiletDetailsPanel';

const FIND_BY_ID = loader('./findLooById.graphql');

const HomePage = ({ initialPosition, ...props }) => {
  const [mapPosition, setMapPosition] = useMapPosition(config.fallbackLocation);

  const { data: loos } = useNearbyLoos({
    variables: {
      lat: mapPosition.center.lat,
      lng: mapPosition.center.lng,
      radius: Math.ceil(mapPosition.radius),
    },
  });

  const selectedLooId = useParams().id;

  const { data, loading } = useQuery(FIND_BY_ID, {
    variables: {
      id: selectedLooId,
      skip: !selectedLooId,
    },
  });

  const isLooPage = useRouteMatch('/loos/:id');

  const [shouldCenter, setShouldCenter] = React.useState(isLooPage);

  // set initial map center to toilet if on /loos/:id
  React.useEffect(() => {
    if (shouldCenter && data) {
      setMapPosition({
        center: data.loo.location,
      });

      // don't recenter the map each time the id changes
      setShouldCenter(false);
    }
  }, [data, shouldCenter, setMapPosition]);

  // set the map position if initialPosition prop exists
  React.useEffect(() => {
    if (initialPosition) {
      setMapPosition({
        center: initialPosition,
      });
    }
  }, [initialPosition, setMapPosition]);

  return (
    <PageLayout onSelectedItemChange={(center) => setMapPosition({ center })}>
      <Box height="100%" display="flex" position="relative">
        <LooMap
          loos={loos.map((loo) => {
            if (loo.id === selectedLooId) {
              return {
                ...loo,
                isHighlighted: true,
              };
            }
            return loo;
          })}
          center={mapPosition.center}
          zoom={mapPosition.zoom}
          onViewportChanged={setMapPosition}
          onSearchSelectedItemChange={setMapPosition}
          showContributor
          showSearchControl
          showLocateControl
        />
        {Boolean(selectedLooId) && (
          <Box
            position="absolute"
            left={0}
            bottom={0}
            width="100%"
            zIndex={100}
          >
            <ToiletDetailsPanel data={data && data.loo} isLoading={loading} />
          </Box>
        )}
      </Box>
    </PageLayout>
  );
};

HomePage.propTypes = {
  initialPosition: PropTypes.shape({
    lat: PropTypes.number,
    lng: PropTypes.number,
  }),
};

export default HomePage;
