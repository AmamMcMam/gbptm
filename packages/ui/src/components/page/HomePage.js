import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { connect } from 'react-redux';
import { Link } from 'react-router';
import MediaQuery from 'react-responsive';
import _ from 'lodash';

import LooMap from '../map/LooMap';
import DismissableBox from '../DismissableBox';
import LooListItem from '../LooListItem';
import Notification from '../Notification';

import styles from '../css/home-page.module.css';
import toiletMap from '../css/loo-map.module.css';
import layout from '../css/layout.module.css';
import headings from '../../css/headings.module.css';
import controls from '../../css/controls.module.css';

import { actionFindNearbyRequest } from '../../redux/modules/loos';
import { actionHighlight } from '../../redux/modules/loo-map-nearest';
import { actionSignoutRequest } from '../../redux/modules/auth';
import { actionToggleViewMode } from '../../redux/modules/app';

import config from '../../config';

export class HomePage extends Component {
  constructor(props) {
    super(props);

    var position = this.props.geolocation.position;

    this.renderList = this.renderList.bind(this);
    this.renderMobileMap = this.renderMobileMap.bind(this);
    this.renderWelcome = this.renderWelcome.bind(this);

    if (!this.props.loos) {
      this.props.actionFindNearbyRequest(
        position.lng,
        position.lat,
        config.nearest.radius
      );
    }
  }

  renderList() {
    var loos = this.props.loos;

    // Loading
    if (!loos) {
      return (
        <Notification>
          <p>Fetching toilets&hellip;</p>
        </Notification>
      );
    }

    // No results
    if (loos && !loos.length) {
      return (
        <Notification>
          <p>No nearby loos found.</p>
        </Notification>
      );
    }

    return (
      <div>
        <h2 className={headings.large}>Nearest Toilets</h2>
        <ul className={styles.looList}>
          {loos &&
            loos.slice(0, config.nearest.limit).map((loo, i) => (
              <li key={i} className={styles.looListItem}>
                <LooListItem
                  loo={loo}
                  onHoverStart={_.partial(this.props.actionHighlight, loo)}
                  onHoverEnd={_.partial(this.props.actionHighlight, undefined)}
                />
              </li>
            ))}
        </ul>
      </div>
    );
  }

  // Rendered as a tab on mobile devices
  renderMobileMap() {
    var loos = this.props.loos;

    return (
      <div className={styles.mobileMap}>
        <div className={toiletMap.map}>
          {!loos && (
            <div className={toiletMap.loading}>Fetching toilets&hellip;</div>
          )}

          <LooMap
            loos={loos}
            initialZoom={14}
            showLocation={true}
            showSearchControl={true}
            showLocateControl={true}
            showCenter={true}
            showFullscreenControl={true}
            shouldCluster={true}
            initialPosition={this.props.geolocation.position}
          />
        </div>
      </div>
    );
  }

  renderWelcome() {
    var content = `
            <p>The ${
              config.nearest.limit
            } nearest toilets are listed below. Click more info to find out about
            each toilet's features.</p><p>You can set preferences to highlight toilets that meet your specific
            needs.</p>
        `;

    return (
      <DismissableBox persistKey="home-welcome" title="Hi!" content={content} />
    );
  }

  render() {
    var mode = this.props.app.viewMode;

    return (
      <div className={styles.container}>
        {/* Logged in message */}
        {this.props.auth.authenticated && (
          <Notification>
            <p>
              Logged in.{' '}
              <a role="button" href onClick={this.props.actionSignoutRequest}>
                Signout
              </a>
            </p>
          </Notification>
        )}

        <div className={layout.controls}>
          {config.allowAddEditLoo && (
            <Link to="/report" className={controls.btn}>
              Add a toilet
            </Link>
          )}

          <MediaQuery maxWidth={config.viewport.mobile}>
            <button
              className={controls.btn}
              onClick={this.props.actionToggleViewMode}
            >
              {mode === 'list' ? 'View map' : 'View list'}
            </button>
          </MediaQuery>
        </div>

        <MediaQuery
          maxWidth={config.viewport.mobile}
          className={styles.mobileContent}
        >
          {mode === 'list' && this.renderWelcome()}
          {mode === 'list' && this.renderList()}
          {mode === 'map' && this.renderMobileMap()}
        </MediaQuery>
        <MediaQuery minWidth={config.viewport.mobile}>
          {this.renderWelcome()}
          {this.renderList()}
        </MediaQuery>
      </div>
    );
  }
}

HomePage.propTypes = {
  loos: PropTypes.array,
};

var mapStateToProps = state => ({
  geolocation: state.geolocation,
  loos: state.loos.nearby,
  auth: state.auth,
  app: state.app,
});

var mapDispatchToProps = {
  actionFindNearbyRequest,
  actionHighlight,
  actionSignoutRequest,
  actionToggleViewMode,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(HomePage);
