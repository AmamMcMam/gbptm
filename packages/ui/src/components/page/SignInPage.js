import React, { Component } from 'react';
import { browserHistory } from 'react-router';
import { connect } from 'react-redux';
import _ from 'lodash';

import { actionAuthRequest } from '../../redux/modules/auth';

import layout from '../css/layout.module.css';
import headings from '../../css/headings.module.css';
import controls from '../../css/controls.module.css';

class SingInPage extends Component {
  render() {
    return (
      <div>
        {this.props.app.canGoBack && (
          <div>
            <div className={layout.controls}>
              <button onClick={browserHistory.goBack} className={controls.btn}>
                Back
              </button>
            </div>
          </div>
        )}

        <h2 className={headings.large}>Hello! What's your name?</h2>

        <p>Before you can contribute data we'll need to know who to thank.</p>
        <p>
          We'll only store the name you give us against the data you contribute.
        </p>
        <p>Use one of the services below to let us know you're real.</p>

        <div className={controls.btnStack}>
          <button
            className={controls.facebook}
            onClick={_.partial(this.props.actionAuthRequest, 'facebook')}
          >
            Facebook
          </button>
          <button
            className={controls.twitter}
            onClick={_.partial(this.props.actionAuthRequest, 'twitter')}
          >
            Twitter
          </button>
          <button
            className={controls.osm}
            onClick={_.partial(this.props.actionAuthRequest, 'openstreetmap')}
          >
            OpenStreetMap
          </button>
          <button
            className={controls.github}
            onClick={_.partial(this.props.actionAuthRequest, 'github')}
          >
            Github
          </button>
        </div>
      </div>
    );
  }
}

var mapStateToProps = state => ({
  app: state.app,
});

var mapDispatchToProps = {
  actionAuthRequest,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(SingInPage);
