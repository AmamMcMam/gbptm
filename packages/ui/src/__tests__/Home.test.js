import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import { HomePage } from '../components/page/HomePage';

import loos from './fixtures/loos.json';

jest.mock('leaflet');

describe('Map', () => {
  var component;
  var defaultProps;

  var actionFindNearbyRequest = jest.fn();
  var actionHighlight = jest.fn();
  var actionSignoutRequest = jest.fn();

  beforeEach(() => {
    defaultProps = {
      geolocation: {
        position: {
          lat: 1.292708,
          lng: 52.6369559,
        },
      },
      loos: loos,
      auth: {
        authenticated: false,
      },
      app: {
        viewMode: 'list',
      },
      actionFindNearbyRequest: actionFindNearbyRequest,
      actionHighlight: actionHighlight,
      actionSignoutRequest: actionSignoutRequest,
    };

    component = <HomePage {...defaultProps} />;
  });

  // Jest snapshot
  it('renders snapshot', () => {
    var tree = renderer.create(component).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders no results', () => {
    var mounted = shallow(<HomePage {...defaultProps} loos={[]} />);

    expect(mounted.contains(<p>No nearby loos found.</p>)).toEqual(true);
  });

  describe('Logout: ', () => {
    it('renders', () => {
      var mounted = shallow(
        <HomePage {...defaultProps} auth={{ authenticated: true }} />
      );

      expect(mounted.find({ onClick: actionSignoutRequest }).length).toBe(1);
    });

    it('hides', () => {
      var mounted = shallow(component);

      expect(mounted.find({ onClick: actionSignoutRequest }).length).toBe(0);
    });
  });
});
