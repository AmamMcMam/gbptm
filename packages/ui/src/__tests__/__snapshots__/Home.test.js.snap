// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map renders snapshot 1`] = `
<div
  className="container"
>
  <div
    className="controls"
  />
</div>
`;
